### PR TITLE
Rewrite js1 week 1 prep content

### PR DESCRIPTION
**Pull request by Dedekind561**
## What does this change?

Module: JS1
Week(s): 1

## Checklist

- [x] I have read the [contributing guidelines](CONTRIBUTING.MD)
- [x] I have checked my spelling and grammar with an [automated tool](https://www.grammarly.com/grammar-check)
- [x] I have previewed my changes to check the [markdown renders](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax) as I intend
- [x] I have run my code to check it works
- [x] My changes follow our [Style Guide](https://curriculum.codeyourfuture.io/guides/code-style-guide)

### CODE REVIEW

File: content/js1/sprints/1/prep/index.md [hunk 1]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+Try predicting what `result` will evaluate to when the code above runs. Try executing this line in the Node REPL to observe what happens.
+
+{{</note>}}
+
+When this program runs, the variable `result` will evaluate to `undefined`.
+
+{{<note title="Note" type="note">}}
+
+`undefined` is a data type in JavaScript often used to represent the absence of a specific value
+
```
Comment (illicitonion) [inline content/js1/sprints/1/prep/index.md:441]: If we added an explanation of what literals are above, I'd add an explanation that the literal value for `undefined` is `undefined` (no quotes), and that `"undefined"` would be a string.

File: content/js1/sprints/1/prep/index.md [hunk 2]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+{{</note>}}
+
+### 🏷️ Saving expressions
+
+In programming we often want to reuse particular expressions.
+Therefore, we need a way to refer back to particular expressions. Let’s consider the following string:
+
+```js
+"Hello there";
```
Comment (illicitonion) [inline content/js1/sprints/1/prep/index.md:224]: I've noticed we're being pretty inconsistent with the use or not use of `;`s in code - do we want to try to be more consistent?

File: content/js1/sprints/1/prep/index.md [hunk 3]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+So this new problem can be expressed as a question:
+
+> ❓ Problem
+>
+> "How can we check the values of variables in our program when it is being executed?"
+
+### 🖨️ Printing to the terminal
+
+In order to check the values when our program runs, we can make use of a function called `console.log`.  
+`console.log` will allow us to log expressions to the console while our program is executing. This is very useful for complex programs when we need to check the value of expressions in our program.
```
Comment (illicitonion) [inline content/js1/sprints/1/prep/index.md:384]: We haven't introduced the term "console" before here. I'd maybe either rephrase this as "`console.log` will allow us to log the values of expressions to the terminal interface while our program is executing", or define that `We often use the term `console` to refer to "the text output of a text interface" - when we're running in a terminal, that means the terminal interface` or something similar.

File: content/js1/sprints/1/prep/index.md [hunk 4]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+
+Let's see how to use `console.log`
+
+In a file called `example.js`, we can write name of the function `console.log`.
+
+```js title="example.js"
+console.log;
+```
+
+However, if we run this file with Node, we won't be able to see anything in the terminal.
```
Comment (illicitonion) [inline content/js1/sprints/1/prep/index.md:394]: I think this block is probably confusing. I'd maybe go the other way around; something like:

````
Just like when we called `Math.round` by adding parentheses with an input value inside them, we call `console.log` in the same way:

```js title="example.js"
console.log("hello there!");
```

Try running the file containing this instruction. You should see the string "hello there!" logged in the terminal.

If you go back to a REPL, and evaluate `console.log`, you'll see it's a function, just like we saw with `Math.round`:

[screenshot]

If we just wrote `console.log` in our file and ran the file, we wouldn't expect to see anything because we're not _calling_ the function. We know that the expression `console.log` evaluates to a function, but until we call it, it doesn't _do_ anything."
````

I'd be tempted to _also_ move the "Logging vs function output" (i.e. side effect vs return value) explanation here, and then use the "Saving function output" section as a "Let's reinforce that" section.

File: content/js1/sprints/1/prep/index.md [hunk 5]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+const result = Math.round(10.3);
+```
+
+When this program is executed, it creates a variable called result and assigns to it **the output of the function**, in this case the rounded number.
+
+So `result` will store the value `10`.
+
+### 🔭 Logging vs function output
+
+Some functions output values that we can use in our program.
```
Comment (illicitonion) [inline content/js1/sprints/1/prep/index.md:420]:
```suggestion
Some functions return values that we can use in our program.
```

I'd like us to avoid using the term "output" when referring to return values - "return" or "evaluate to" are great, but "output" is super ambiguous and I think pretty unclear when we're trying to distinguish what is output to the terminal vs what is "output" by a function

File: content/js1/sprints/1/prep/index.md (hunk 3 above)
Comment (illicitonion) [inline content/js1/sprints/1/prep/index.md:384]: We haven't introduced the term "console" before here. I'd maybe either rephrase this as "`console.log` will allow us to log the values of expressions to the terminal interface while our program is executing", or define that `We often use the term `console` to refer to "the text output of a text interface" - when we're running in a terminal, that means the terminal interface` or something similar.

File: content/js1/sprints/1/prep/index.md [hunk 6]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+
+However, some functions don't produce useful outputs in our running program; however they can still cause **effects**. Consider the following line:
+
+```js
+const result = console.log("hello world");
+```
+
+{{<note title="Exercise" type="exercise">}}
+
+Try predicting what `result` will evaluate to when the code above runs. Try executing this line in the Node REPL to observe what happens.
```
Comment (illicitonion) [inline content/js1/sprints/1/prep/index.md:432]:
```suggestion
Try predicting what `result` will evaluate to when the code above runs. Try executing this line in the Node REPL and then evaluating the value of the `result` variable to observe what happens.
```

File: content/js1/sprints/1/prep/index.md [hunk 7]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+
+When this program runs, the variable `result` will evaluate to `undefined`.
+
+{{<note title="Note" type="note">}}
+
+`undefined` is a data type in JavaScript often used to represent the absence of a specific value
+
+{{</note>}}
+
+This is potentially confusing as `console.log` is a function with a set of instructions. `console.log` _does_ have an effect: it logs values to the console. However, `console.log` doesn't produce an output that we can use _inside_ the rest of our running program
```
Comment (illicitonion) [inline content/js1/sprints/1/prep/index.md:444]: I would include a screenshot here of evaluating these two lines:
```console
> const result = console.log("hello world");
hello world
undefined
> result
undefined
> console.log(result)
undefined
undefined
```

and highlighting that the first and last evaluates result in _two_ lines of different colours, the side-effect print and the value evaluated to.

I'd maybe also go back to where we were introducing the REPL and include a similar screenshot just showing:

```console
> 5 + 10
15
```

and highlighting that `15` is the value evaluated by the expression, and maybe even reprise it here, so we have a consistent frame of reference to show the "there's an extra console.log output line" point.

I find that our trainees frequently get very confused between "what the REPL says a value evaluated to" and "what was console.log'd" because the difference is super subtle.

File: content/js1/sprints/1/prep/index.md [hunk 8]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+
+{{<note type="info">}}
+
+Key fact: `console.log` is used to print values to the terminal. It doesn’t produce an output in the running program.
+
+{{</note>}}
```
Comment (illicitonion) [inline content/js1/sprints/1/prep/index.md:450]: I feel like this ends very bluntly; I'd be tempted to add a wrap-up/recap section reprising the learning objectives here.

**Review by illicitonion — COMMENTED** [review]
I've left a bunch of comments here for slightly more "open ended discussion" things, and filed #159 against this branch with a bunch of edits - very happy to discuss anything/everything!

File: content/js1/sprints/1/prep/index.md [hunk 9]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
 
+## Prerequisites
 
+- Basic arithmetic
+- Access to a Linux/MacOS terminal
+
+## Learning objectives
+
+- Define a terminal
+- Describe the purpose of VSCode
```
Comment (illicitonion) [inline content/js1/sprints/1/prep/index.md:19]: We don't use or talk about VSCode at all in this text.

File: content/js1/sprints/1/prep/index.md [hunk 10]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+
+{{<note title="Activity" type="info">}}
+
+Try and define the **user interface** for the following devices:
+
+- a calculator
+- a microwave
+- a desktop lamp
+- Alexa
+- ChatGPT
```
Comment (40thieves) [inline content/js1/sprints/1/prep/index.md:77]: Might be interesting to introduce a website here? (I suppose ChatGPT is, but a different kettle of fish).

I was thinking that Facebook might be an interesting example?

File: content/js1/sprints/1/prep/index.md [hunk 11]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+
+## Notes
+
+### 💻 Interacting with computers
+
+Modern computers are highly complex devices: itt would be too difficult and time-consuming to list all the components that make up a modern computer. To begin understanding, we can start with a simple definition of a computer:
+
+{{<note title="Definition" type="info">}}
+
+A computer is a device used to store and process data. Computers have the capacity to store data through their memory and the ability to change this data through their processing ability.
```
Comment (40thieves) [inline content/js1/sprints/1/prep/index.md:50]: We're using slightly inconsistent language here vs the "Writing computer instructions" section: here we refer to "processing" and there we refer to "operations".

Might be nice to make those consistent?

File: content/js1/sprints/1/prep/index.md [hunk 12]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+
+So we can think of a programming language as allowing us to create **data** and **operations** on that data.
+
+{{<note title="Definition" type="info">}}
+**Data** consists of information in different forms: for example, numbers and text could be different forms of data. The data in a executing program is sometimes referred to as the **state**.
+{{</note>}}
+
+{{<note title="Definition" type="info">}}
+**Operations** are anything that take data and modify existing data or create new data from the current data in the program.
+{{</note>}}
```
Comment (40thieves) [inline content/js1/sprints/1/prep/index.md:115]: I'm not sure this section flows very well? I think I'd restructure to something like this:

We can issue commands to the computer using the terminal. These commands are instructions that the computer knows how to interpret. Given the ls command, the computer knows this is an instruction to list the files and directories in the current directory.

During the execution of a computer program, a computer will store and data and modify it using operations

Definitions of data & operations...

`ls` in the terminal is an example of a programming language.

[Definition] A programming language is a particular set of rules for writing computer instructions.

We can use other programming languages too. In this course we're going to focus on the JavaScript programming language.

File: content/js1/sprints/1/prep/index.md [hunk 13]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+A programming language will have certain ways of organising data. They split data up into different categories called **data types**.
+
+{{<note title="Definition" type="info">}}
+A **data type** is a grouping of data with some particular properties.
+{{</note>}}
+
+In JavaScript, we have several different data types, like **numbers** and **strings**.
+
+For example, `10` is part of the number data type.  
+`3.14` is also part of the number data type, as we include both integers non-integers in this data type.
```
Comment (40thieves) [inline content/js1/sprints/1/prep/index.md:130]:
```suggestion
`3.14` is also part of the number data type, as we include both integers and non-integers in this data type.
```

File: content/js1/sprints/1/prep/index.md [hunk 14]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+
+A programming language will have certain ways of organising data. They split data up into different categories called **data types**.
+
+{{<note title="Definition" type="info">}}
+A **data type** is a grouping of data with some particular properties.
+{{</note>}}
+
+In JavaScript, we have several different data types, like **numbers** and **strings**.
+
+For example, `10` is part of the number data type.
```
Comment (40thieves) [inline content/js1/sprints/1/prep/index.md:129]: I think I would switch this to the less formal form as I think it's more approachable.

```suggestion
`10` is an example of the number data type.
```

File: content/js1/sprints/1/prep/index.md [hunk 15]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+b) What is the **product** of 10 and 32?  
+c) What is the **remainder** when 32 is divided by 10?
+
+From question a), instead of saying "the sum of 10 and 32" we can rewrite this description using numbers and a symbol. So “the sum of 10 and 32” can be written as `10 + 32` as follows:
+
+```raw title="symbolic form"
+10 + 32
+```
+
+In symbolic form, `+` is an operator: an **operator** is a symbol that is used to represent some kind of operation. In this example, `+` is used to represent the add operation “add the 2 numbers together”.
```
[the remark this answers is not in the document]
Comment (40thieves) [reply content/js1/sprints/1/prep/index.md:149]: Agreed with your points, but I'd go further to suggest "make the sum of the numbers" instead (i.e. removing the reference to another partially related number)

File: content/js1/sprints/1/prep/index.md [hunk 16]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+
+## Key vocabulary
+
+terminal, console, function, call, invoke, argument, parentheses, REPL, evaluate, implementation opaque
+
+## Notes
+
+### 💻 Interacting with computers
+
+Modern computers are highly complex devices: itt would be too difficult and time-consuming to list all the components that make up a modern computer. To begin understanding, we can start with a simple definition of a computer:
```
[the remark this answers is not in the document]
Comment (40thieves) [reply content/js1/sprints/1/prep/index.md:46]: Heh, it seems I suggested the exact opposite in the previous incarnation of this PR: https://github.com/CodeYourFuture/syllabus/pull/659#discussion_r1155126884

File: content/js1/sprints/1/prep/index.md [hunk 17]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+`Math.round(10.3)` is a call expression: it can be thought of as saying:
+
+"apply the set of instructions for `Math.round` to the number `10.3`."
+
+If we type `Math.round(10.3)` then again we get the result 10.
+So we say that `Math.round(10.3)` outputs 10.
+
+{{<note title="Note" type="note">}}
+It's important to note that `Math.round` is **implementation opaque**.
+**Implementation opaque** means we can’t read the set of instructions for `Math.round`. This is because `Math.round` was authored by the developers that created NodeJS.
```
Comment (40thieves) [inline content/js1/sprints/1/prep/index.md:332]: This feels like we're emphasising something that we actually don't want trainees to know lots of depth - in other words, knowing what "implementation opaque" means is not relevant to getting a job.

I think the intention is to say "don't worry about the set of instructions for `Math.round`". But I think a lot of trainees would write down "implementation opaque" and try to memorise it as an important fact.

Could we just swap it out for something like:

```suggestion
We can't read the set of instructions for `Math.round` because it was written by the developers of NodeJS itself.
```

File: content/js1/sprints/1/prep/index.md [hunk 18]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+
+So far we’ve seen how expressions can be evaluated using the Node REPL. The Node REPL is a very useful tool for evaluating expressions quickly.
+
+However, most of the time, we need to write more complex programs in files that consist of many lines.
+Instead of using the REPL mode, we can get Node to execute the instructions written in our file.
+
+We can use the node command to run a JavaScript file.
+A JavaScript file ends with `.js` - this is known as the file extension.
+
+Let’s suppose we have a file `sums.js` like this:
```
Comment (40thieves) [inline content/js1/sprints/1/prep/index.md:345]: Given that the file contains operation other than sums, could we rename this file to something like `arithmetic.js`?

File: content/js1/sprints/1/prep/index.md [hunk 19]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+```
+
+In this case, we can write the command `node sums.js`.
+
+This command is an instruction to execute the program written inside index.js. Our program consists of 3 lines, each line an expression. So the machine will start:
+
+`10 + 3;` the computer will evaluate this expression  
+`10 * 3;` the computer will evaluate this expression  
+`10 / 3;` the computer will evaluate this expression
+
```
Comment (40thieves) [inline content/js1/sprints/1/prep/index.md:360]: Not sure where to put this comment, but there's a relatively big gap between exercises here - I wonder if we can come up with something interactive for variables and/or function calling?

File: content/js1/sprints/1/prep/index.md [hunk 20]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+```js title="sums.js"
+10 + 3;
+10 * 3;
+10 / 3;
+```
+
+In this case, we can write the command `node sums.js`.
+
+This command is an instruction to execute the program written inside index.js. Our program consists of 3 lines, each line an expression. So the machine will start:
```
Comment (40thieves) [inline content/js1/sprints/1/prep/index.md:355]: Perhaps we could change this for consistency?

```suggestion
This command is an instruction to execute the program written inside index.js. Our program consists of 3 lines, each line an expression. So the computer will start:
```

File: content/js1/sprints/1/prep/index.md [hunk 21]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+
+We can use variables to store the output from a function.
+
+Let’s consider the following file, `arithmetic.js`;
+
+```js title="arithmetic.js"
+const result = Math.round(10.3);
+```
+
+When this program is executed, it creates a variable called result and assigns to it **the output of the function**, in this case the rounded number.
```
Comment (40thieves) [inline content/js1/sprints/1/prep/index.md:414]: As mentioned in the original PR, [I'd like to call out](https://github.com/CodeYourFuture/syllabus/pull/659/files#r1155142553) that calling a function is itself an expression, the value of which is then assigned to the variable.

**Review by 40thieves — COMMENTED** [review]
Generally looking really good so much better than what we had before, thanks @Dedekind561!

I think there some points that I'd like to see considered from [the original PR to the syllabus repo](https://github.com/CodeYourFuture/syllabus/pull/659), specifically I would like to see the points about motivation around expressions and how function calls are expressions.

File: content/js1/sprints/1/prep/index.md [hunk 22]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+
+So `result` will store the value `10`.
```
Comment (carl0FF) [inline content/js1/sprints/1/prep/index.md:416]: This sentence sounds weird to me, but it could be just because it's not my first language. I would rewrite it to something like 
> So the value of `result` will be `10`

or 
> So the value `10` will be stored in `result`

Feel free to ignore this comment though

File: content/js1/sprints/1/prep/index.md [hunk 23]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+
+```js title="example.js"
+console.log("hello there!");
+```
+
+We should see the string "hello there!" logged out in the terminal.
+
+### Saving function output
+
+We can use variables to store the output from a function.
```
Comment (carl0FF) [inline content/js1/sprints/1/prep/index.md:406]: In line with @40thieves comment below, this could be rewritten to something like:

> Since function calls are also expressions, their output can also be stored in variables, same as the math expressions we've seen earlier.

**Comment by carl0FF** [comment]
This is just brilliant, so much effort and care into it! 👏

File: content/js1/sprints/1/prep/index.md [hunk 24]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+
+Once we have the name for a function, we need to get Node to read the function's instructions and execute them. Execution means the computer reads the instructions and carries them out. So we can write the following in the REPL:
+
+![round-img](math.round-number-10.3.png)
+
+Notice now we have `(` and `)` parentheses after the name of the function and a number inside the parentheses. The parentheses indicate that we are calling the function. The number inside the parentheses is the **input** we're passing to the function.
+
+{{<note title="Definition" type="info">}}
+
+**Calling a function** means telling the computer to read the function's instructions and carry out its instructions. When calling a function we can also pass inputs to the function.
```
Comment (carl0FF) [inline content/js1/sprints/1/prep/index.md:319]: Would it make sense to introduce here the synonyms for "calling a function"? Invoking a function, executing a function... any others?

File: content/js1/sprints/1/prep/index.md (hunk 9 above)
Comment (Dedekind561) [reply content/js1/sprints/1/prep/index.md:19]: VSCode is mentioned in the Git resources at the start of the course, so I don't think we need this learning objective here

File: content/js1/sprints/1/prep/index.md (hunk 2 above)
Comment (Dedekind561) [reply content/js1/sprints/1/prep/index.md:224]: Yeah I'll add some additional `;` for consistency throughout

File: content/js1/sprints/1/prep/index.md (hunk 5 above)
Comment (Dedekind561) [reply content/js1/sprints/1/prep/index.md:469]: I think `return` is the clearest: let's stick with that for now. I don't think pairing `input` and `return` will be much of an issue.

File: content/js1/sprints/1/prep/index.md [hunk 25]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+### Reusing instructions
+
+Let’s consider another scenario. Instead of adding or multiplying numbers, we’ll consider a number like `10.3`.  
+Given the number `10.3`, we could ask:
+
+> 🤔 "What is the nearest whole number to `10.3`?"
+
+The process of finding the nearest whole number to 10.3 is called **rounding**. So we could rephrase the previous question to be:
+
+> 🤔 "What does the number `10.3` **round** to?”
```
[the remark this answers is not in the document]
Comment (Dedekind561) [reply content/js1/sprints/1/prep/index.md:309]: Just spending more time on rounding here because we're using it as the foothold for introducing functions.

File: content/js1/sprints/1/prep/index.md (hunk 17 above)
Comment (Dedekind561) [reply content/js1/sprints/1/prep/index.md:377]: Cool - I'm going to remove the idea from this point in the curriculum and refer to it in week 2.

File: content/js1/sprints/1/prep/index.md [hunk 26]
```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
...
+console.log("hello there!");
+```
+
+We should see the string "hello there!" logged out in the terminal.
+
+### Saving function output
+
+We can use variables to store the output from a function.
+
+Let’s consider the following file, `arithmetic.js`;
```
[the remark this answers is not in the document]
Comment (Dedekind561) [reply content/js1/sprints/1/prep/index.md:453]: It's just referring to a file with some code in it, I'm going to update this to something like: "Let's suppose we have the file..."

File: content/js1/sprints/1/prep/index.md (hunk 7 above)
Comment (Dedekind561) [reply content/js1/sprints/1/prep/index.md:501]: My explanation was getting really long here towards the end here: I'm going to move this idea into a challenge and question in the coursework.

File: content/js1/sprints/1/prep/ls-feedback.png
[the remark this answers is not in the document]
Comment (Dedekind561) [reply content/js1/sprints/1/prep/ls-feedback.png:1]: Good spot: I don't think so, but I'll redo the screenshot

File: content/js1/sprints/1/prep/index.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (Dedekind561) [reply content/js1/sprints/1/prep/index.md]: I'm going to leave out the term abstraction here for the time being. We can return to it in week 4, though in more depth when we discuss APIs.

File: content/js1/sprints/1/prep/index.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Dedekind561) [reply content/js1/sprints/1/prep/index.md]: I've added those definitions in.
